File: algorithmic_toolbox/week6_dynamic_programming2/partition3.py
def n_partition(n, total, elements):
    if total % n != 0:
        return 0
    count = n
    while count:
        can_partition, elements = partition(total//n, elements)
        # print(can_partition, elements)
        if not can_partition:
            return 0
        count -= 1
    return 1


def partition(value, elements):
    dp = []
    for _ in range(0, value+1):
        dp.append([False for _ in range(len(elements)+1)])

    for i in range(1, value+1):
        for j in range(1, len(elements)+1):
            if dp[i][j-1] or i == elements[j-1]:
                dp[i][j] = True
            else:
                value_diff = i - elements[j-1]
                if value_diff > 0 and dp[value_diff][j-1]:
                    dp[i][j] = True
    
    selected_idx = set()
    temp_value = value
    for j in range(1, len(elements)+1):
        while temp_value and j and dp[temp_value][j]:
            if not dp[temp_value][j-1]:
                selected_idx.add(j-1)
                temp_value -= elements[j-1]
            j -= 1
    
    new_elements = []
    for j in range(len(elements)):
        if j not in selected_idx:
            new_elements.append(elements[j])

    return dp[value][len(elements)], new_elements


def k_partition(nums, k):
	target, m = divmod(sum(nums), k)
	if m: return False
	dp, n = [0]*k, len(nums)
	nums.sort(reverse=True)
	def dfs(i):
		if i == n:
			return len(set(dp)) == 1
		for j in range(k):
			dp[j] += nums[i]
			if dp[j] <= target and dfs(i+1):
				return True
			dp[j] -= nums[i]
			if not dp[j]: break
		return False
	return nums[0] <= target and dfs(0)


# [4,4,6,2,3,8,10,2,10,7]
# 4
if __name__ == '__main__':
    input()
    elements = [int(x) for x in input().split(' ')]
    # elements.sort()
    # print(n_partition(3, sum(elements), elements[::-1]))
    can_partition = k_partition(elements, 3)
    if can_partition:
        print(1)
    else:
        print(0)

File: algorithmic_toolbox/week6_dynamic_programming2/partition3.go
package main

import (
	"fmt"
	"sort"
)

func nPartition(n int, total int, eles []int) bool {
	sort.Slice(eles, func(i, j int) bool {
		return eles[i] > eles[j]
	})
	canPartition := false
	partitionCount := n
	if total%n != 0 {
		return false
	}
	for partitionCount > 0 {
		canPartition, eles = partition(total / n, eles)
		// fmt.Println(canPartition, eles)
		if !canPartition {
			return false
		}
		partitionCount--
	}
	return true
}

// case example: 9 & 7 2 2 2 2 2 2 2 3, res: wrong
// case example: 6 & 2 1 4 4 4 6, res: wrong
// 4 [4,4,6,2,3,8,10,2,10,7], res: correct
// 2 partition ref: http://thisthread.blogspot.com/2018/02/2-partition-problem.html
func partition(value int, eles []int) (bool, []int) {
	// init DP 2D array
	dp := make([][]bool, value+1)
	for i := 0; i <= value; i++ {
		dp[i] = make([]bool, len(eles)+1)
	}
	for i := 1; i <= value; i++ {
		for j := 1; j <= len(eles); j++ {
			if dp[i][j-1] || i == eles[j-1] {
				dp[i][j] = true
			} else {
				valueDiff := i - eles[j-1]
				if valueDiff > 0 && dp[valueDiff][j-1] {
					dp[i][j] = true
				}
			}
		}
		// fmt.Println(dp[i])
	}

	// rule out selected values
	selectedIdx := make(map[int]bool)
	tmpVal := value
	for j := 1; j <= len(eles); j++ {
		for tmpVal >= 0 && j >= 1 && dp[tmpVal][j] {
			if !dp[tmpVal][j-1] {
				selectedIdx[j-1] = true
				tmpVal -= eles[j-1]
			}
			j--
		}
	}
	// fmt.Println(selectedIdx)
	newEles := []int{}
	for i, ele := range eles {
		if !selectedIdx[i] {
			newEles = append(newEles, ele)
		}
	}
	return dp[value][len(eles)], newEles
}

func main() {
	var n, total int
	fmt.Scanln(&n)
	eles := []int{}

	for i := 0; i < n; i++ {
		var ele int
		fmt.Scanf("%d", &ele)
		total += ele
		eles = append(eles, ele)
	}
    res := nPartition(3, total, eles)
	// partition(total/3, eles)
	if res {
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
}

// func main() {
//     s := []int{8, 2, 6, 3, 1, 4} 
//     sort.Slice(s, func(i, j int) bool {
//         return s[i] > s[j]
//     })
//     fmt.Println(s)  // [8 6 4 3 2 1]
// }
